ipset.go

```diff
@@ -15,25 +18,24 @@ import "sort"
 // nothing on an empty set. Ranges may be fully, partially, or not
 // overlapping.
 type IPSet struct {
+	mu sync.Mutex
+
 	// in are the ranges in the set.
 	in []IPRange
 
 	// out are the ranges to be removed from 'in'.
 	out []IPRange
-}
 
-// toInOnly updates s to clear s.out, by merging any s.out into s.in.
-func (s *IPSet) toInOnly() {
-	if len(s.out) > 0 {
-		s.in = s.Ranges()
-		s.out = nil
-	}
+	// normalized indicates that 'in' is in sorted order, and 'out' is
+	// empty.
+	normalized bool
```

Comment:
How big of a performance win does this provide? Keeping this extra bit of state around. Could we calculate it on the fly cheaply when we need it?

Reply:
Didn't measure, it's possible this is pointless. Could make ranges smarter to detect an already-sorted s.in and empty s.out, so it can fast-exit after a single pass through s.in. Probably better, I'll alter to do that.

Reply:
Removed normalized, and instead added a mild fast path to rangesLocked that avoids allocating if it can. With that, rangesLocked still costs a slice sort plus one pass through the array, but not an alloc+copy unless removals are required.